File: android/app/src/main/java/com/crispycrew/fryday/widget/SharedFileStorage.kt
package com.crispycrew.fryday.widget

import android.content.Context
import java.io.File

// filesDir 안의 파일을 atomic 하게 read/write.
// SharedPreferences 의 in-memory 캐시 stale 이슈 회피용.
// 파일 시스템은 커널 매개 → 어느 컴포넌트(위젯 receiver / app) 에서 read 하든 최신 값 보장.
object SharedFileStorage {
    const val PENDING_FILE = "widget-pending.json"
    const val TODOS_FILE = "widget-todos.json"
    const val STATE_FILE = "widget-state.json"

    // 프로세스 내 read-modify-write race 방지용 lock (Mutex 대체 - 동기 lock)
    private val lock = Any()

    private fun file(context: Context, name: String): File {
        return File(context.filesDir, name)
    }

    fun readString(context: Context, name: String): String? {
        synchronized(lock) {
            val f = file(context, name)
            if (!f.exists()) return null
            return try {
                f.readText(Charsets.UTF_8)
            } catch (_: Exception) {
                null
            }
        }
    }

    fun writeString(context: Context, name: String, contents: String): Boolean {
        synchronized(lock) {
            val target = file(context, name)
            val tmp = File(context.filesDir, "$name.tmp")
            return try {
                tmp.writeText(contents, Charsets.UTF_8)
                // atomic rename
                if (target.exists()) target.delete()
                tmp.renameTo(target)
            } catch (_: Exception) {
                tmp.delete()
                false
            }
        }
    }

    fun delete(context: Context, name: String): Boolean {
        synchronized(lock) {
            val f = file(context, name)
            return if (f.exists()) f.delete() else true
        }
    }
}

File: targets/widget/SharedFileStorage.swift
import Foundation

// App Group 컨테이너의 파일을 NSFileCoordinator 로 read/write.
// UserDefaults 크로스프로세스 sync 이슈 회피용.
enum SharedFileStorage {
    static let appGroupID = "group.com.fryday.shared"

    static let pendingFile = "widget-pending.json"
    static let todosFile = "widget-todos.json"
    static let stateFile = "widget-state.json"

    // 위젯 → 앱 파일 변경 알림 (Darwin notification, 크로스프로세스)
    static let pendingChangedNotification = "com.fryday.widget-pending-changed" as CFString

    static func postPendingChangedNotification() {
        CFNotificationCenterPostNotification(
            CFNotificationCenterGetDarwinNotifyCenter(),
            CFNotificationName(pendingChangedNotification),
            nil, nil, true
        )
    }

    private static func url(for fileName: String) -> URL? {
        FileManager.default
            .containerURL(forSecurityApplicationGroupIdentifier: appGroupID)?
            .appendingPathComponent(fileName)
    }

    static func readString(_ fileName: String) -> String? {
        guard let fileURL = url(for: fileName) else { return nil }
        var readError: NSError?
        var result: String? = nil
        let coordinator = NSFileCoordinator()
        coordinator.coordinate(readingItemAt: fileURL, options: [], error: &readError) { url in
            result = try? String(contentsOf: url, encoding: .utf8)
        }
        return result
    }

    @discardableResult
    static func writeString(_ contents: String, to fileName: String) -> Bool {
        guard let fileURL = url(for: fileName) else { return false }
        var writeError: NSError?
        var success = false
        let coordinator = NSFileCoordinator()
        coordinator.coordinate(writingItemAt: fileURL, options: .forReplacing, error: &writeError) { url in
            do {
                try contents.write(to: url, atomically: true, encoding: .utf8)
                success = true
            } catch {
                success = false
            }
        }
        return success
    }

    static func readJSON<T: Decodable>(_ fileName: String, as type: T.Type) -> T? {
        guard let str = readString(fileName),
              let data = str.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }

    static func readJSONObject(_ fileName: String) -> Any? {
        guard let str = readString(fileName),
              let data = str.data(using: .utf8) else { return nil }
        return try? JSONSerialization.jsonObject(with: data)
    }

    @discardableResult
    static func writeJSON<T: Encodable>(_ value: T, to fileName: String) -> Bool {
        guard let data = try? JSONEncoder().encode(value),
              let str = String(data: data, encoding: .utf8) else { return false }
        return writeString(str, to: fileName)
    }
}

struct PendingState: Codable {
    var ids: [String]
    var before: [String: Bool]

    static let empty = PendingState(ids: [], before: [:])
}
